src/Tables.jl

```diff
@@ -148,34 +216,89 @@ end
     if @generated
-        vals = Tuple(:(allocatecolumn($(fieldtype(NT, i)), len)) for i = 1:fieldcount(NT))
+        vals = Tuple(:(allocatecolumn($(fieldtype(types, i)), len)) for i = 1:fieldcount(types))
         return :(NamedTuple{names}(($(vals...),)))
     else
-        return NamedTuple{names}(Tuple(allocatecolumn(fieldtype(NT, i), len) for i = 1:fieldcount(NT)))
+        return NamedTuple{names}(Tuple(allocatecolumn(fieldtype(types, i), len) for i = 1:fieldcount(types)))
     end
 end
 
+haslength(T) = T === Base.HasLength() || T === Base.HasShape{1}()
+
 # add! will push! or setindex! a value depending on if the row-iterator HasLength or not
-@inline add!(val, col::Int, nm::Symbol, ::Base.HasLength, nt, row) = setindex!(nt[col], val, row)
+@inline add!(val, col::Int, nm::Symbol, ::Union{Base.HasLength, Base.HasShape{1}}, nt, row) = setindex!(nt[col], val, row)
 @inline add!(val, col::Int, nm::Symbol, T, nt, row) = push!(nt[col], val)
 
-@inline function buildcolumns(sch, rowitr::T) where {T}
+@inline function buildcolumns(schema, rowitr::T) where {T}
     L = Base.IteratorSize(T)
-    len = L == Base.HasLength() ? length(rowitr) : 0
-    nt = allocatecolumns(sch, len)
+    len = haslength(L) ? length(rowitr) : 0
+    nt = allocatecolumns(schema, len)
     for (i, row) in enumerate(rowitr)
-        unroll(add!, sch, row, L, nt, i)
+        eachcolumn(add!, schema, row, L, nt, i)
     end
     return nt
 end
 
+@inline function add_or_widen!(val::T, col::Int, nm::Symbol, L, columns, row, allocate, len) where {T}
+    if !allocate
+        @inbounds columns[col] = add_or_widen!(columns[col], val, L, row)
+    else
+        @inbounds columns[col] = add_or_widen!(allocatecolumn(T, len), val, L, row)
+    end
+    return
+end
+
+@inline function add_or_widen!(dest::AbstractVector{T}, val::S, ::Union{Base.HasLength, Base.HasShape{1}}, row) where {T, S}
+    if S === T || val isa T
+        @inbounds dest[row] = val
+        return dest
+    else
+        new = allocatecolumn(Base.promote_typejoin(T, S), length(dest))
+        copyto!(new, 1, dest, 1, row - 1)
+        @inbounds new[row] = val
+        return new
+    end
+end
+
+@inline function add_or_widen!(dest::AbstractVector{T}, val::S, ::Base.SizeUnknown, row) where {T, S}
+    if S === T || val isa T
+        push!(dest, val)
+        return dest
+    else
+        new = allocatecolumn(Base.promote_typejoin(T, S), length(dest))
+        copyto!(new, dest)
+        push!(new, el)
+        return new
+    end
+end
+
+# when Tables.types(x) === nothing
+function buildcolumns(sch::Schema{names, nothing}, rowitr::T) where {names, T}
+    state = iterate(rowitr)
+    state === nothing && return NamedTuple{names}(Tuple(Missing[] for _ in names))
+    row::eltype(rowitr), st = state
+    cols = length(names)
+    L = Base.IteratorSize(T)
+    len = haslength(L) ? length(rowitr) : 0
+    columns = Vector{AbstractVector}(undef, cols)
```

Comment:
One of the things that got me stuck when trying to implement this in this framework is that it's not completely clear what structure should contain the columns. `NamedTuple` is out of the question because they can change type, but doesn't a `Vector{AbstractVector}` cause performance issues?

Reply:
I was even thinking, maybe the only performant way to do this is to use generated functions that create code where we use a different variable per each column, but maybe that's overkill.

Reply:
Good point. Indeed, since the loop over columns is applied for each row, it's essential to avoid any type uncertainty.

Maybe we can avoid a generated function though, and instead use the same strategy as `map`: store columns as a named tuple, and have the function call itself with the updated named tuple each time the types change.

Reply:
That's the strategy in IndexedTables. However here it's a bit tricky to implement for technical reasons as in IndexedTables you are iterating over `NamedTuples` so you can tell "statically" whether you need to expand or not, just by looking at the type of the row. One strategy I was proposing [here](https://github.com/JuliaData/Tables.jl/issues/7#issuecomment-417658109) is to store columns as a `NamedTuple`, and change `add!` so that if the types are incompatible, it would return `nothing` and not do anything. Then in turn `eachcolumns` would know that if a function call returns `nothing`, it has to exit and signal that it failed, so the main loop would just call `eachcolumns` and only call itself with the updated named tuple when `eachcolumn` signals that it failed.